modules/cluster-management/src/main/java/org/apache/ignite/internal/cluster/management/ClusterManagementGroupManager.java

```diff
@@ -487,27 +467,46 @@ private CompletableFuture<Void> joinCluster(CmgRaftService service, ClusterTag c
      * Starts the CMG Raft service using the provided node names as its peers.
      */
     private CompletableFuture<CmgRaftService> startCmgRaftService(Collection<String> nodeNames) {
-        // TODO: wait for nodes to appear, see https://issues.apache.org/jira/browse/IGNITE-16811
-        List<ClusterNode> nodes = resolveNodes(clusterService, nodeNames);
-
         try {
             return raftManager
                     .prepareRaftGroup(
                             CMG_RAFT_GROUP_NAME,
-                            nodes,
-                            () -> {
-                                clusterStateStorage.start();
-
-                                return new CmgRaftGroupListener(clusterStateStorage);
-                            },
+                            resolveNodes(clusterService, nodeNames),
+                            this::createCmgRaftGroupListener,
+                            this::createCmgRaftGroupEventsListener,
                             RaftGroupOptions.defaults()
                     )
                     .thenApply(service -> new CmgRaftService(service, clusterService));
-        } catch (NodeStoppingException e) {
+        } catch (Exception e) {
             return failedFuture(e);
         }
     }
 
+    private CmgRaftGroupListener createCmgRaftGroupListener() {
+        clusterStateStorage.start();
```

Comment:
It is pretty mean to hide a side effect inside a method which looks like a 'pure' method if you look at its name :) How about renaming the method? Although all the candidate names seem to be too long.

Reply:
Renamed to `startCmgRaftGroupListener`

Reply:
I've inlined this method